extensions/wrapper/clients/typescript-client/src/oauth2/AccessTokenStore.ts

```diff
@@ -0,0 +1,38 @@
+import {AccessTokenFetcher} from './AccessTokenFetcher';
+
+export class AccessTokenStore {
+    _accessToken?: string;
+    accessTokenFetcher: AccessTokenFetcher;
+    private isRefreshing = false;
+    private currentFetchPromise: Promise<string | undefined> =
+        Promise.resolve(undefined);
+
+    constructor(accessTokenFetcher: AccessTokenFetcher, token?: string) {
+        this.accessToken = token;
+        this.accessTokenFetcher = accessTokenFetcher;
+    }
+
+    get accessToken(): string | undefined {
+        return this._accessToken;
+    }
+
+    set accessToken(token: string | undefined) {
+        this._accessToken = token;
+    }
+
+    // Synchronized refreshing of the access token
+    async refreshToken() {
+        if (this.isRefreshing) {
+            await this.currentFetchPromise;
+            return;
+        }
+
+        this.isRefreshing = true;
+        this.currentFetchPromise = this.accessTokenFetcher.fetch();
+        const newToken = await this.currentFetchPromise;
+        if (newToken) {
+            this.accessToken = newToken;
+        }
+        this.isRefreshing = false;
```

Comment:
Comment: Theoretically it would be enough to have:

```ts
activeRequest: Promise<string> | null = null;
//...
async refreshToken() {
   if (this.activeRequest) {
       return this.activeRequest;
   }
   this.activeRequest = this.accessTokenFetcher.fetch()
       .then((token) => this.accessToken = token);
       .finally(() => this.activeRequest = null));
   return this.activeRequest;
}
```